Fill.py

```diff
@@ -129,18 +142,35 @@ def distribute_items_restrictive(world: MultiWorld, fill_locations=None):
         else:
             restitempool.append(item)
 
-    world.random.shuffle(fill_locations)
-    call_all(world, "fill_hook", progitempool, nonexcludeditempool, localrestitempool, nonlocalrestitempool, restitempool, fill_locations)
+    call_all(world, "fill_hook", progitempool, nonexcludeditempool,
+             localrestitempool, nonlocalrestitempool, restitempool, fill_locations)
+
+    locationDeficit = len(progitempool) - len(prioritizedlocations)
+    if locationDeficit > 0:
+        if locationDeficit > len(defaultlocations):
+            raise FillError(
+                f'Not enough locations for progress items. There are {len(progitempool)} progression items with {len(prioritizedlocations)} priority locations and {len(defaultlocations)} default locations')
+        for i in range(0, locationDeficit):
```

Comment:
range operations like prioritizedlocations = defaultlocations[:locationDeficit] are much faster

Reply:
Updated to use array slices instead